chips/cc26x0/src/udma.rs

```diff
@@ -2,42 +2,72 @@
 /// # Micro Direct Memory Access for the TI CC26x0 Microcontroller
 ///
 
-use kernel;
 use kernel::common::regs::{ReadOnly, ReadWrite, WriteOnly};
+use kernel::ReturnCode;
+use prcm;
 
 pub const UDMA_BASE: usize = 0x4002_0000;
-pub const UDMA_CONFIG_BASE: usize = 0x2000_0400;
 
 #[repr(align(1024))]
 pub struct ControlTable {
-    config_array: [u8; 1024],
+    config_array: [DMAConfig; 32]
 }
 
-/*
+#[derive(Copy, Clone)]
 struct DMAConfig {
     source_ptr: usize,
     dest_ptr: usize,
-    control: usize,
+    control: ReadWrite<u32,DMATableControl::Register>,
     _unused: usize,
 }
-*/
 
-pub static mut DMACTRLTAB: ControlTable = ControlTable::new();
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone, PartialEq)]
+#[repr(u8)]
+pub enum DMAPeripheral {
+    SOFTWARE_0 = 0,
+    UART0_RX = 1,
+    UART0_TX = 2,
+    SSP0_RX = 3,
+    SSP0_TX = 4,
+    AUX_ADC = 7,
+    AUX_SW = 8,
+    GPT0_A = 9,
+    GPT0_B = 10,
+    GPT1_A = 11,
+    GPT1_B = 12,
+    AON_PROG2 = 13,
+    DMA_PROG = 14,
+    AON_RTC = 15,
+    SSP1_RX = 16,
+    SSP1_TX = 17,
+    SOFTWARE_1 = 18,
+    SOFTWARE_2 = 19,
+    SOFTWARE_3 = 20,
+}
+
+
+static mut DMACTRLTAB: ControlTable = ControlTable::new();
 
 impl ControlTable {
-    pub const fn new() -> ControlTable {
+    const fn new() -> ControlTable {
         ControlTable {
-            config_array: [0; 1024]
+            config_array: [DMAConfig {
```

Comment:
another option is to replace this array initialization with some kind of manual initialization where 32 separate instances of DMAConfig are cast from 4 byte blocks of memory. This also doesn't seem like the right thing to do.

Reply:
Oh yeah, this is a thing in Rust. https://github.com/rust-lang/rfcs/issues/1109

None doesn't implement copy either, so it's why the process array needs to be:
```
static mut PROCESSES: [Option<&'static mut kernel::Process<'static>>; NUM_PROCS] = [
    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
    None, None, None, None,
];
```

A general macro-based solution here would solve both of the problems. This answer claims to do it (https://stackoverflow.com/questions/28656387/initialize-a-large-fixed-size-array-with-non-copy-types/28666474#28666474) but I'm not sure if it's what we want.

Reply:
@brghena exactly- I was balking at the mem::uninitialized call (the compiler gave me all sort of red flags)

maybe the process array approach is the best idea- each channel needs to be enabled anyway. I'll try it out and report back!